docs/design-proposals/etcd/ETCD-Data-Partition-Design.md

```diff
@@ -51,17 +54,156 @@ number, watch from revision number, etc.
 It will allow maximum 8K events per millisecond in a single cluster, up to 64 clusters, till 
 approximately January 1st, 2500.
 
+### ETCD Objects Partition
+
+* There will be a default ETCD cluster called system partition. It has all the data that is not
+created/owned by [tenants](../multi-tenancy/multi-tenancy-overview.md). Hence, it will have system
+objects such as nodes, as well as tenant objects as they are created and managed by system admin.
+* There will have multiple ETCD clusters called tenant partitions. They have all the data that is
+created/owned by tenants, including namespaces for a tenant, pods created in one of the namespaces
+of the tenants, etc. A ETCD tenant cluster can have objects from multiple tenants. Objects for a single
+tenant can only located in one tenant cluster.
+
+#### API Changes
+
+* We added a top level Kubernetes object named '**TenantStorage**' with path '**/api/v1/tenantstorages**'.
+In addition, we added a field named '**StorageClusterId**' into '**[Tenant](../multi-tenancy/multi-tenancy-api-resource-model.md)**' object.
+  - This allows admin to view and allocate tenants to etcd clusters manually
+  - This allows api server to discover tenant clusters and decide where to create/locate objects
+
+```yaml
+apiVersion: v1
+kind: TenantStorage 
+storageClusterId: "c1"
+serviceAddress: "172.30.0.122:2379"
+metadata:
+  name: "tenant-cluster-1"
+```
+
+```yaml
+apiVersion: v1
+kind: Tenant
+spec:
+  storageClusterId: c1
```

Comment:
It looks like storageClusterId is unique. Could we move it from spec to metadata?

Reply:
Name validation is across all objects. Would like StorageClusterId to be capable of having separate validation rule.

Reply:
In that case we might need additional indexers.